### PR TITLE
Add ability to attach, detach and attach once again

### PR DESCRIPTION
**Pull request by DonJayamanne**
@int19h @ericsnowcurrently @karthiknadig 
**The code is far from complete.**, but would like some early feedback on the solution. 
With the changes I've made I've managed to get re-attach working, at least on VSC.

Pending - Fix remainder of the code and what ever tests fail

Summary:
* Create new threading to check for connections
* Change how callbacks are passed into pydevd

### CODE REVIEW

File: ptvsd/wrapper.py [hunk 1]
```diff
@@ -45,9 +45,10 @@
 __version__ = "4.0.0a5"
 
 
-#def ipcjson_trace(s):
-#    print(s)
-#ipcjson._TRACE = ipcjson_trace
+def ipcjson_trace(s):
+    with open('/Users/donjayamanne/Desktop/Development/vscode/my_forked_ptvsd/ptvsd/log.log', 'a') as file:
```
Comment (DonJayamanne) [inline ptvsd/wrapper.py]: Temporary debugging code 😃

File: ptvsd/wrapper.py [hunk 2]
```diff
@@ -1674,23 +1695,22 @@ def _new_sock():
...
-
-    server_thread = threading.Thread(target=proc.process_messages,
-                                     name=name)
+    server_thread = threading.Thread(target=proc.process_messages, name=name)
     server_thread.daemon = True
     server_thread.start()
 
     if addhandlers:
         _add_atexit_handler(proc, server_thread)
         _set_signal_handlers(proc)
```
Comment (DonJayamanne) [inline ptvsd/wrapper.py]: I'm going to have to modify `_set_signal_handlers` to close all VS Code message processors.

File: ptvsd/wrapper.py [hunk 3]
```diff
@@ -1726,6 +1746,14 @@ def start_server(port, addhandlers=True):
     server = _create_server(port)
     client, _ = server.accept()
     pydevd = _start(client, server)
+    def _wait_for_more():
+        while True:
+            client, _ = server.accept()
+            _add_socket_handler(client, pydevd, 'ptvsd.Server', killonclose=True, addhandlers=False)
+
+    connection_thread = threading.Thread(target=_wait_for_more, name='ptvsd_client_connection')
```
Comment (DonJayamanne) [inline ptvsd/wrapper.py]: New thread that'll ensure we support more than one client.
@int19h When speaking to Karthik the question about limiting multiple concurrent client connections came up.
Should we have a limit on the number of concurrent connections, to just one?

Comment (int19h) [reply ptvsd/wrapper.py]: Yes. We never supported anything more than that before, and just thinking of all the possibilities of things going wrong in that scenario makes my head hurt. :)

File: ptvsd/wrapper.py [hunk 4]
```diff
@@ -1726,6 +1775,16 @@ def start_server(port, addhandlers=True):
...
     client, _ = server.accept()
     pydevd = _start(client, server)
+
+    def _wait_for_more_connections():
+        while True:
+            client, _ = server.accept()
+            _add_pydevd_event_handler(client, pydevd, name='ptvsd.Server')
+
+    connection_thread = threading.Thread(target=_wait_for_more_connections,
+                                         name='ptvsd_client_connection')
```
Comment (karthiknadig) [inline ptvsd/wrapper.py]: Use `ptvsd.Client.Connection` as name, to be consistent with other `ptvsd` thread names

File: ptvsd/wrapper.py [hunk 5]
```diff
@@ -1779,6 +1838,7 @@ def start_client(host, port, addhandlers=True):
     """
     client = _create_client()
     client.connect((host, port))
+    pydevd = PydevdSocket()
```
Comment (karthiknadig) [inline ptvsd/wrapper.py]: That is not needed, right?

Comment (DonJayamanne) [reply ptvsd/wrapper.py]: 👍

**Comment by DonJayamanne** [comment]
@ericsnowcurrently 
Here are the requirements for remote debugging:
* Current functionality of PR
* Run a program
    - Somewhere down the line, we can import ptvsd and start the socket server (PTVSD)
    - It should listen for connections (**non-blocking**)
        This is a scenario where we have the socket server listening for connections, but in a non-blocking manner.
    - If a connection is made, then debugger should connect

**Comment by DonJayamanne** [comment]
Closing in favor of #296

**Pull request closed without merging**